backends/utils/history_server.go

```diff
@@ -0,0 +1,526 @@
+			i++
+			rowsInCurrentDF++
+			nullColumns = append(nullColumns, &currentNullColumns)
+		}
+
+		if lineIterator.Error() != nil {
+			return lineIterator.Error()
+		}
+	}
+
+	if iter.Err() != nil {
+		return iter.Err()
+	}
+
+	if len(indexColData) > 0 {
+		frame, err := createDF(userNameColData, backendNameColData, tableNameColData, actionTypeColData, containerColData,
+			actionDurationColData, startTimeColData,
+			indexColData, customColumnsByName, nullColumns, hasNullsInCurrentDF)
+		if err != nil {
+			return err
+		}
+
+		out <- frame
+	}
+
+	return nil
+}
+
+func createDF(userNameColData, backendNameColData, tableNameColData, actionTypeColData, containerColData []string,
+	actionDurationColData []int64, startTimeColData []time.Time,
+	indexColData []int, customColumnsByName map[string][]string, nullColumns []*pb.NullValuesMap, hasNullsInCurrentDF bool) (frames.Frame, error) {
+	var columns []frames.Column
+	col, _ := frames.NewSliceColumn("UserName", userNameColData)
+	columns = append(columns, col)
+	col, _ = frames.NewSliceColumn("BackendName", backendNameColData)
+	columns = append(columns, col)
+	col, _ = frames.NewSliceColumn("TableName", tableNameColData)
+	columns = append(columns, col)
+	col, _ = frames.NewSliceColumn("ActionDuration", actionDurationColData)
+	columns = append(columns, col)
+	col, _ = frames.NewSliceColumn("StartTime", startTimeColData)
+	columns = append(columns, col)
+	col, _ = frames.NewSliceColumn("ActionType", actionTypeColData)
+	columns = append(columns, col)
+	col, _ = frames.NewSliceColumn("Container", containerColData)
+	columns = append(columns, col)
+
+	indexColumn, _ := frames.NewSliceColumn("id", indexColData)
+
+	for colName, colData := range customColumnsByName {
+		col, _ = frames.NewSliceColumn(colName, colData)
+		columns = append(columns, col)
+	}
+
+	nullMask := nullColumns
+	if !hasNullsInCurrentDF {
+		nullMask = nil
+	}
+
+	frame, err := frames.NewFrameWithNullValues(columns, []frames.Column{indexColumn}, nil, nullMask)
+	if err != nil {
+		return nil, err
+	}
+
+	return frame, nil
+}
+
+func (m *HistoryServer) writeMonitoringBatch(logs []HistoryEntry) {
+	var data []byte
+
+	for _, log := range logs {
+		d, err := json.Marshal(log)
+		if err != nil {
+			m.logger.Warn("Failed to marshal logs to json, err: %v. logs: %v", err, logs)
+			return
+		}
+
+		data = append(data, d...)
+		data = append(data, []byte("\n")...)
+	}
+
+	logFilePath := m.getCurrentLogFileName()
+	input := &v3io.PutObjectInput{Path: logFilePath, Body: data, Append: true}
+	err := m.container.PutObjectSync(input)
+
+	if err != nil {
+		// retry on 5xx errors
+		if errWithStatusCode, ok := err.(v3ioerrors.ErrorWithStatusCode); ok && errWithStatusCode.StatusCode() > 500 {
+			input := &v3io.PutObjectInput{Path: logFilePath, Body: data, Append: true}
+			err = m.container.PutObjectSync(input)
+		}
+
+		if err != nil {
+			m.logger.Error("Failed to append Frames History logs to file, err: %v. logs: %v", err, logs)
```

Comment:
Here too, only log the log we failed on, not all of them. Not sure why history is capitalized 🤓

Reply:
it writes the whole batch of logs, so the whole request failed